core/src/main/java/com/datastax/oss/driver/internal/core/protocol/BytesToSegmentDecoder.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.protocol;
+
+import com.datastax.oss.driver.api.core.connection.CrcMismatchException;
+import com.datastax.oss.protocol.internal.Segment;
+import com.datastax.oss.protocol.internal.SegmentCodec;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
+import java.nio.ByteOrder;
+import net.jcip.annotations.NotThreadSafe;
+
+/**
+ * Decodes {@link Segment}s from a stream of bytes.
+ *
+ * <p>This works like a regular length-field-based decoder, but we override {@link
+ * #getUnadjustedFrameLength} to handle two peculiarities: the length is encoded on 17 bits, and we
+ * also want to check the header CRC before we use it. So we parse the whole segment header ahead of
+ * time, and store it until we're ready to build the segment.
+ */
+@NotThreadSafe
+public class BytesToSegmentDecoder extends LengthFieldBasedFrameDecoder {
+
+  private final SegmentCodec<ByteBuf> segmentCodec;
+  private SegmentCodec.Header header;
+
+  public BytesToSegmentDecoder(@NonNull SegmentCodec<ByteBuf> segmentCodec) {
+    super(
+        // max length (Netty wants this to be the overall length including everything):
+        segmentCodec.headerLength()
+            + SegmentCodec.CRC24_LENGTH
+            + Segment.MAX_PAYLOAD_LENGTH
+            + SegmentCodec.CRC32_LENGTH,
+        // offset and size of the "length" field: that's the whole header
+        0,
+        segmentCodec.headerLength() + SegmentCodec.CRC24_LENGTH,
```

Comment:
why only `segmentCodec.headerLength() + SegmentCodec.CRC24_LENGTH,` and not 
`segmentCodec.headerLength() + SegmentCodec.CRC24_LENGTH  + Segment.MAX_PAYLOAD_LENGTH  + SegmentCodec.CRC32_LENGTH`
?

Reply:
Short answer: it's the value that makes the tests pass :trollface: 
Long answer: it's how many bytes the parent class needs to read before it can invoke `getUnadjustedFrameLength`. What we do in there is `segmentCodec.decodeHeader`, which just needs the header and header CRC.